src/gax-internal/src/observability/grpc_tracing.rs

```diff
@@ -114,10 +114,75 @@ where
     fn call(&mut self, req: http::Request<B>) -> Self::Future {
         let attempt_count = req.extensions().get::<AttemptCount>().map(|a| a.0);
         let span = create_grpc_span(req.uri(), &self.layer.inner, attempt_count);
-        Box::pin(self.inner.call(req).instrument(span))
+        let future = self.inner.call(req);
+        Box::pin(ResponseFuture {
+            inner: future,
+            span,
+        })
     }
 }
 
+/// A future that wraps the inner service's future and records the status code on completion.
+#[pin_project::pin_project]
+pub struct ResponseFuture<F> {
+    #[pin]
+    inner: F,
+    span: tracing::Span,
+}
+
+impl<F, ResBody, Error> Future for ResponseFuture<F>
+where
+    F: Future<Output = Result<http::Response<ResBody>, Error>>,
+    Error: std::fmt::Display,
+{
+    type Output = Result<http::Response<ResBody>, Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let _guard = this.span.enter();
+        let result = futures_util::ready!(this.inner.poll(cx));
```

Comment:
This works, but I had to read the code of `futures_util::ready!` to convince myself.  Maybe with more experience I will not care, but it is not obvious that this "leaves" the span when the inner future is not ready.

Maybe a small comment?